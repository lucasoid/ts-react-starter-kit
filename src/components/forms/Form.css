.ctrl,
.btnGroup {
    margin-top: 1em;
}

.ctrl input[type='text'] {
    width: 100%;
    padding: 0.25em;
}

.ctrl label {
    display: block;
    padding: 0.25em 0;
}

.btnGroup button {
    margin-right: 1em;
}
