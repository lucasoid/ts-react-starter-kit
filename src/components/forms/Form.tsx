import * as React from 'react';
import { Button, ButtonTypes } from '~components/ui/Button';
import * as styles from './Form.css';

interface IFormProps {
    onSubmit?: Function;
    onCancel?: Function;
    submitText?: string;
    cancelText?: string;
    children?: any;
}

export const Form: React.FC<IFormProps> = props => {
    return (
        <form
            onSubmit={e => {
                e.preventDefault();
                if (props.onSubmit !== undefined) props.onSubmit();
            }}
        >
            {props.children}
            <div className={styles.btnGroup}>
                {props.onCancel !== undefined && (
                    <Button type={ButtonTypes.SECONDARY} onClick={props.onCancel}>
                        {props.cancelText}
                    </Button>
                )}
                {props.onSubmit !== undefined && (
                    <Button type={ButtonTypes.PRIMARY} onClick={props.onSubmit}>
                        {props.submitText}
                    </Button>
                )}
            </div>
        </form>
    );
};
