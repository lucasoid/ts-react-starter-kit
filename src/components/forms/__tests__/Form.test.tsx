import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { Form } from '../Form';

describe('Form', () => {
    it('renders a form element conditionally with save and cancel buttons', () => {
        const onSubmit = jest.fn();
        const onCancel = jest.fn();

        // if onSubmit and onCancel are undefined, the buttons are not shown
        let wrapper = shallow(<Form>Form content here</Form>);
        expect(wrapper).toMatchSnapshot();

        wrapper = shallow(
            <Form onSubmit={onSubmit} onCancel={onCancel}>
                Form content here
            </Form>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders customized button text', () => {
        const props = {
            onSubmit: jest.fn(),
            onCancel: jest.fn(),
            submitText: 'My custom submit text',
            cancelText: 'My custom cancel text',
        };
        let wrapper = shallow(<Form {...props}>Form content here</Form>);
        expect(wrapper).toMatchSnapshot();
    });

    it('invokes the onSubmit prop when the form is submitted', () => {
        const onSubmit = jest.fn();
        let wrapper = mount(<Form onSubmit={onSubmit} />);
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
        expect(onSubmit.mock.calls.length).toBe(1);
    });

    it('invokes the onCancel prop when the cancel button is clicked', () => {
        const onCancel = jest.fn();
        let wrapper = mount(<Form onCancel={onCancel} />);
        wrapper.find('button').simulate('click', { preventDefault: () => {} });
        expect(onCancel.mock.calls.length).toBe(1);
    });

    it('gracefully handles a missing onSubmit prop', () => {
        let wrapper = mount(<Form />);
        expect(() => wrapper.find('form').simulate('submit', { preventDefault: () => {} })).not.toThrow();
    });
});
