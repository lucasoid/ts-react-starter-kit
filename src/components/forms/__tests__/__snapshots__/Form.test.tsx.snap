// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders a form element conditionally with save and cancel buttons 1`] = `
<form
  onSubmit={[Function]}
>
  Form content here
  <div
    className="btnGroup"
  />
</form>
`;

exports[`Form renders a form element conditionally with save and cancel buttons 2`] = `
<form
  onSubmit={[Function]}
>
  Form content here
  <div
    className="btnGroup"
  >
    <Button
      onClick={[MockFunction]}
      type={1}
    />
    <Button
      onClick={[MockFunction]}
      type={0}
    />
  </div>
</form>
`;

exports[`Form renders customized button text 1`] = `
<form
  onSubmit={[Function]}
>
  Form content here
  <div
    className="btnGroup"
  >
    <Button
      onClick={[MockFunction]}
      type={1}
    >
      My custom cancel text
    </Button>
    <Button
      onClick={[MockFunction]}
      type={0}
    >
      My custom submit text
    </Button>
  </div>
</form>
`;
